naas/client.py

```diff
@@ -39,18 +39,44 @@ def default_accept_header(cls):
         return header_value
 
     @classmethod
-    def get(cls, url, headers={}, params={}):
+    def get(cls, url, headers=None, params=None):
         """Execute a GET request"""
+        if headers is None:
+            headers = {}
+        if params is None:
+            params = {}
         request_headers = {**Client.default_headers(), **headers}
-        return requests.get(url, headers=request_headers)
+        return requests.get(url, headers=request_headers, params=params)
 
     @classmethod
-    def head(cls, url, headers={}, params={}):
+    def head(cls, url, headers=None, params=None):
         """Execute a HEAD request"""
+        if headers is None:
+            headers = {}
+        if params is None:
+            params = {}
         request_headers = {**Client.default_headers(), **headers}
-        return requests.head(url, headers=request_headers)
+        return requests.head(url, headers=request_headers, params=params)
 
     @classmethod
-    def routes(cls):
-        """Routes"""
-        return Client.get(Client.api_host())
+    def post(cls, url, headers=None, json=None):
+        if headers is None:
+            headers = {}
+        if json is None:
+            json = {}
+        request_headers = {**Client.default_headers(), **headers}
+        return requests.post(url, headers=request_headers, json=json)
```

Comment:
@muhallan - The recommendation I would make here is not to assume `json`. For instance, we may need to take in other MIME types (IE: CSV, Excel, etc), which would be a different `Content-Type` and `Accept` header. The `POST` body can be any one of those things, so it may seem confusing if the client says `json`.

Reply:
Thank you. I am modifying this to include an option to send both files and additional data.

Reply:
@muhallan - I stand corrected. The `requests` library takes in a `json` argument, as well as `data` and `files`. According to their docs:

> Instead of encoding the `dict` yourself, you can also pass it directly using the `json` parameter and it will be encoded automatically. Note, the `json` parameter is ignored if either `data` or `files` is passed.

I really don't link this overloading and precedence, and would _prefer_ only having the `data`. 

When testing I noticed I was getting some errors:

```
>>> naas.requests.Campaigns.create_by_project_id('tmnt', params={'name': 'testing', 'description': 'My description'})
<Response [500]>
```

This was because the parameters that were sent to the API look like:

```ruby
{"campaign"=>"description", "project_id"=>"tmnt"}
```

However, if I manually set the `Content-Type` header to `application/json` and used `json.dumps` it would work as expected:

```
>>> Client.post(url, headers={'Content-Type': 'application/json' }, data=json.dumps(request_body))
<Response [201]>
```

Which sent:

```ruby
{"campaign"=>{"name"=>"Testing", "description"=>"My Description"}, "project_id"=>"tmnt"}
```

We may want to fix this up within the `requests` or the `client` itself to properly send `JSON` in these requests, using the HTTP content negotiation.

Let me know if there are any questions here.

Reply:
Let me look through this and get fixes. Thanks.

Reply:
Sounds good @muhallan

Reply:
Can I do this
`Client.post(url, headers={'Content-Type': 'application/json' }, data=json.dumps(request_body))`
for all post requests? I know we might need attachments. How can I tell which ones should need attachments?

I have found out that if I use `data` keyword, then I need to encode it as `json` and then set the header like that.